### PR TITLE
Fix macOS smartctl IOKit paths in ndsudo

### LINKED ISSUE #23444: [Bug]: go.d smartctl collector can never run on macOS — ndsudo rejects @, which is mandatory in every IOKit device path
### Bug description

Netdata version: 2.10.4 (stable), macOS build-from-source install.

Environment: macOS, Apple Silicon and Intel both affected -- this isn't hardware-specific, it's a property of how macOS names storage devices.

ndsudo's argument validator (check_string() in src/collectors/utils/ndsudo.c) only permits: A-Z a-z 0-9 space _ - / . , : =

Every macOS IOKit registry path uses @ for node addressing at every level of the tree (e.g. apciec3@30000000, pci-bridge@0, NS_01@1), and smartctl --scan on macOS always returns device identifiers in exactly this form -- there is no alternate /dev/sdX-style name to fall back to. As a result, go.d's smartctl-json-device-info call via ndsudo fails validation for every device on every Mac, unconditionally:

```
level=warn msg="failed to get device info for 'IOService:/AppleARMPE/.../IONVMeBlockStorageDevice@1' type 'nvme': ... execution failed: ... exit status 2 (stderr: invalid characters in parameters: command line argument No 3 includes invalid character '@')" plugin=go.d collector=smartctl job=test
```

No smartctl job can ever be created on macOS via auto-discovery or manual config as a result -- this isn't an edge case, it's a complete blocker for the entire collector on the entire platform.

### Expected behavior

ndsudo should accept the characters that appear in macOS's own device-naming scheme for arguments it is asked to pass through to smartctl, since it cannot function on macOS at all otherwise.

### Steps to reproduce

1. On any Mac with smartmontools installed: smartctl --scan -- note every path is IOService:/.../Node@Address/...
2. Start netdata 2.10.4+ with the smartctl collector enabled (default).
3. Check collector.log for the "invalid characters in parameters" warning above, once per discovered device.
4. Reproduce directly, bypassing go.d: sudo /usr/local/netdata/usr/libexec/netdata/plugins.d/ndsudo smartctl-json-device-info --deviceName "IOService:/AppleARMPE/arm-io/AppleT600xIO/ans@8F400000/AppleASCWrapV4/iop-ans-nub/RTBuddy(ANS2)/RTBuddyService/AppleANS3CGv2Controller/NS_01@1" --deviceType nvme --powerMode standby -- fails: invalid characters in parameters: command line argument No 3 includes invalid character '@'
5. Confirm smartctl itself has no issue with this exact path when invoked directly (outside ndsudo): sudo smartctl -a -d nvme "IOService:/AppleARMPE/arm-io/AppleT600xIO/ans@8F400000/AppleASCWrapV4/iop-ans-nub/RTBuddy(ANS2)/RTBuddyService/AppleANS3CGv2Controller/NS_01@1" -- works fine, SMART data returned.

### Installation method

kickstart.sh

### System info

```shell
Darwin <hostname> 25.6.0 Darwin Kernel Version 25.6.0: Sat Jul 11 15:26:21 PDT 2026; root:xnu-12377.161.13~4/RELEASE_ARM64_T6000 arm64
ProductName:		macOS
ProductVersion:		26.6.1
BuildVersion:		25G76
```

### Netdata build info

```shell
Packaging:
    Netdata Version ____________________________________________ : v2.10.4
    Installation Type __________________________________________ : kickstart-build
    Package Architecture _______________________________________ : unknown
    Package Distro _____________________________________________ : unknown
    Configure Options __________________________________________ : cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_C_STANDARD=11 -DCMAKE_CXX_STANDARD=17 -DBUILD_SHARED_LIBS=OFF -DCMAKE_C_FLAGS=' -fexceptions -fno-omit-frame-pointer -funwind-tables -fasynchronous-unwind-tables' -DCMAKE_CXX_FLAGS='  -fexceptions -fno-omit-frame-pointer -funwind-tables -fasynchronous-unwind-tables' -DCMAKE_COMPILE_DEFINITIONS='' -DCMAKE_EXE_LINKER_FLAGS=' -fexceptions -rdynamic' -DCMAKE_SHARED_LINKER_FLAGS=''
Default Directories:
    User Configurations ________________________________________ : /usr/local/netdata/etc/netdata
    Stock Configurations _______________________________________ : /usr/local/netdata/usr/lib/netdata/conf.d
    Ephemeral Databases (metrics data, metadata) _______________ : /usr/local/netdata/var/cache/netdata
    Permanent Databases ________________________________________ : /usr/local/netdata/var/lib/netdata
    Plugins ____________________________________________________ : /usr/local/netdata/usr/libexec/netdata/plugins.d
    Static Web Files ___________________________________________ : /usr/local/netdata/usr/share/netdata/web
    Log Files __________________________________________________ : /usr/local/netdata/var/log/netdata
    Lock Files _________________________________________________ : /usr/local/netdata/var/lib/netdata/lock
    Home _______________________________________________________ : /usr/local/netdata/var/lib/netdata
Operating System:
    Kernel _____________________________________________________ : Darwin
    Kernel Version _____________________________________________ : 25.6.0
    Operating System ___________________________________________ : macOS
    Operating System ID ________________________________________ : macOS
    Operating System ID Like ___________________________________ : macOS
    Operating System Version ___________________________________ : 26.6.1
    Operating System Version ID ________________________________ : none
    Detection __________________________________________________ : sw_vers
Hardware:
    CPU Cores __________________________________________________ : 10
    CPU Frequency ______________________________________________ : unknown
    RAM Bytes __________________________________________________ : 34359738368
    Disk Capacity ______________________________________________ : 494384795648
    CPU Architecture ___________________________________________ : arm64
    Virtualization Technology __________________________________ : unknown
    Virtualization Detection ___________________________________ : none
Container:
    Container __________________________________________________ : unknown
    Container Detection ________________________________________ : none
    Container Orchestrator _____________________________________ : none
    Container Operating System _________________________________ : none
    Container Operating System ID ______________________________ : none
    Container Operating System ID Like _________________________ : none
    Container Operating System Version _________________________ : none
    Container Operating System Version ID ______________________ : none
    Container Operating System Detection _______________________ : none
Features:
    Built For __________________________________________________ : MacOS
    Netdata Cloud ______________________________________________ : YES
    Health (trigger alerts and send notifications) _____________ : YES
    Streaming (stream metrics to parent Netdata servers) _______ : YES
    Back-filling (of higher database tiers) ____________________ : YES
    Replication (fill the gaps of parent Netdata servers) ______ : YES
    Streaming and Replication Compression ______________________ : YES (zstd lz4 gzip brotli)
    Contexts (index all active and archived metrics) ___________ : YES
    Tiering (multiple dbs with different metrics resolution) ___ : YES (5)
    Machine Learning ___________________________________________ : YES
    Memory Allocator ___________________________________________ : system
Database Engines:
    dbengine (compression) _____________________________________ : YES (zstd lz4)
    alloc ______________________________________________________ : YES
    ram ________________________________________________________ : YES
    none _______________________________________________________ : YES
Connectivity Capabilities:
    ACLK (Agent-Cloud Link: MQTT over WebSockets over TLS) _____ : YES
    static (Netdata internal web server) _______________________ : YES
    WebRTC (experimental) ______________________________________ : NO
    Native HTTPS (TLS Support) _________________________________ : YES
    TLS Host Verification ______________________________________ : YES
Libraries:
    LZ4 (extremely fast lossless compression algorithm) ________ : YES
    ZSTD (fast, lossless compression algorithm) ________________ : YES
    zlib (lossless data-compression library) ___________________ : YES
    Brotli (generic-purpose lossless compression algorithm) ____ : YES
    protobuf (platform-neutral data serialization protocol) ____ : YES (system)
    OpenSSL (cryptography) _____________________________________ : YES
    libdatachannel (stand-alone WebRTC data channels) __________ : NO
    JSON-C (lightweight JSON manipulation) _____________________ : YES
    libcap (Linux capabilities system operations) ______________ : NO
    libcrypto (cryptographic functions) ________________________ : YES
    libyaml (library for parsing and emitting YAML) ____________ : YES
    libmnl (library for working with netfilter) ________________ : NO
    stacktraces (library for getting stack traces) _____________ : unknown
Plugins:
    apps (monitor processes) ___________________________________ : YES
    cgroups (monitor containers and VMs) _______________________ : NO
    cgroup-network (associate interfaces to CGROUPS) ___________ : NO
    proc (monitor Linux systems) _______________________________ : NO
    tc (monitor Linux network QoS) _____________________________ : NO
    diskspace (monitor Linux mount points) _____________________ : NO
    freebsd (monitor FreeBSD systems) __________________________ : NO
    macos (monitor MacOS systems) ______________________________ : YES
    windows (monitor Windows systems) __________________________ : NO
    statsd (collect custom application metrics) ________________ : YES
    timex (check system clock synchronization) _________________ : YES
    idlejitter (check system latency and jitter) _______________ : YES
    bash (support shell data collection jobs - charts.d) _______ : YES
    debugfs (kernel debugging metrics) _________________________ : NO
    cups (monitor printers and print jobs) _____________________ : YES
    ebpf (monitor system calls) ________________________________ : NO
    freeipmi (monitor enterprise server H/W) ___________________ : NO
    network-viewer (monitor TCP/UDP IPv4/6 sockets) ____________ : NO
    systemd-journal (monitor journal logs) _____________________ : NO
    windows-events (monitor Windows events) ____________________ : NO
    nfacct (gather netfilter accounting) _______________________ : NO
    perf (collect kernel performance events) ___________________ : NO
    slabinfo (monitor kernel object caching) ___________________ : NO
    Xen ________________________________________________________ : NO
    Xen VBD Error Tracking _____________________________________ : NO
Exporters:
    AWS Kinesis ________________________________________________ : NO
    GCP PubSub _________________________________________________ : NO
    MongoDB ____________________________________________________ : NO
    Prometheus (OpenMetrics) Exporter __________________________ : YES
    Prometheus Remote Write ____________________________________ : NO
    Graphite ___________________________________________________ : YES
    Graphite HTTP / HTTPS ______________________________________ : YES
    JSON _______________________________________________________ : YES
    JSON HTTP / HTTPS __________________________________________ : YES
    OpenTSDB ___________________________________________________ : YES
    OpenTSDB HTTP / HTTPS ______________________________________ : YES
    All Metrics API ____________________________________________ : YES
    Shell (use metrics in shell scripts) _______________________ : YES
Debug/Developer Features:
    Trace All Netdata Allocations (with charts) ________________ : NO
    Developer Mode (more runtime checks, slower) _______________ : NO
Runtime Information:
    Profile ____________________________________________________ : standalone
    Stream Parent (accept data from Children) __________________ : NO
    Stream Child (send data to a Parent) _______________________ : NO
    Total System Memory ________________________________________ : 34359738368
    Available System Memory ____________________________________ : 13555318784
```

### Additional info

Root cause:

check_string() in src/collectors/utils/ndsudo.c hardcodes an allowlist that doesn't include @. It also doesn't include ( or ) -- and real node names on this hardware include parentheses (RTBuddy(ANS2), seen directly in smartctl --scan output), so fixing only @ will likely leave a subset of devices still broken.

Suggested fix:

Add @, (, ) to the allowlist in check_string(), or relax the check specifically for arguments matching the ^IOService: prefix, so other command types aren't affected.

Related: this is filed alongside a separate, unrelated bug in the native macos.plugin nvme_smart collector (netdata/netdata#23443) that surfaced during the same investigation -- distinct subsystem, distinct root cause, noted here only for cross-reference.

### PR DESCRIPTION
**Pull request by ktsaou**
Fixes #23444.

## Summary

- permit `@`, `(`, and `)` only in bounded `IOService:/` values used by the Apple `smartctl-json-device-info --deviceName` contract
- keep the global `ndsudo` argument allowlist unchanged for every other command and value
- handle a bare `ndsudo --test` invocation without reading past `argv`
- add macOS CI coverage against the installed setuid helper, including exact smartctl argv validation

The issue is valid for default smartctl autodiscovery because smartctl returns canonical IOKit paths that the helper rejected. Short manually configured names such as `disk0` and `/dev/disk0` already passed the existing validator, so the fix remains scoped to scanner-produced `IOService:/` paths.

## Validation

- strict Apple and forced non-Apple compilation with `-Wall -Wextra -Werror`
- full Netdata build
- positive 511-byte and negative 512-byte boundary checks
- command, variable, prefix, duplicate, punctuation, missing-value, and malformed-invocation checks
- normal execution as the `netdata` user through a `4750 root:netdata` helper, asserting the exact argv recorded by smartctl
- sanitized real-device read through the rebuilt helper

---
## Summary by cubic
Fixes `ndsudo` on macOS to accept canonical IOKit `IOService:/...` paths in `smartctl` calls, so autodiscovered devices no longer get rejected. Adds macOS CI coverage against the installed setuid helper and ensures `ndsudo` is installed in native macOS builds.

- **Bug Fixes**
  - Allow `@`, `(`, `)` in `--deviceName` for `smartctl-json-device-info` when the value starts with `IOService:/`, enforcing a 511-byte max path.
  - Keep the global `ndsudo` argument allowlist unchanged for all other commands and values.
  - Handle a bare `--test` safely and update help text to document allowed characters.

- **Dependencies**
  - Always install `ndsudo` on macOS via CMake (sets `NEED_NDSUDO` on macOS).
  - macOS CI installs `jq` and `smartmontools`, runs `tests/ndsudo-macos.sh` against the installed helper; build step disables `go` and `scripts` plugins in CI.

<sup>Written for commit 0f75b07831774ffde8f6e4955b5e370e4301082c. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/netdata/netdata/pull/23447?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved macOS device-path validation for storage health checks.
  * Added clearer handling for missing, duplicate, or invalid command-line arguments.
  * Prevented help and test options from being misinterpreted during command processing.
  * Enabled required macOS support for device-information commands.

* **Tests**
  * Added comprehensive macOS validation coverage, including boundary-length paths and expected command failures.
  * Added automated checks for smartctl output and argument handling.
  * Expanded macOS build validation with required tooling.

### CODE REVIEW

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -1311,11 +1311,11 @@ jobs:
         if: needs.file-check.outputs.run == 'true'
         with:
           submodules: recursive
-      - name: Install latest bash
+      - name: Install latest bash and smartmontools
         id: install-bash
         if: needs.file-check.outputs.run == 'true'
         run: |
-          brew install bash
+          brew install bash smartmontools
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply .github/workflows/build.yml]: Valid finding. jq and smartmontools are now declared in a dedicated dependency step that runs only on macos-15-M1, matching the one matrix entry that executes the integration harness. The missing installed helper was addressed separately by making ndsudo a native macOS CMake dependency; the macOS installer invocation now explicitly disables Go and scripts, so the fix does not expand those three jobs into Go builds.

**Comment by ktsaou** [comment]
@cubic-dev-ai please review again

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -1325,7 +1329,12 @@ jobs:
         id: build-source
         if: needs.file-check.outputs.run == 'true'
         run: |
-          sudo bash ./netdata-installer.sh --install-no-prefix /usr/local/netdata --dont-wait --dont-start-it --one-time-build --enable-plugin-otel
+          sudo bash ./netdata-installer.sh --install-no-prefix /usr/local/netdata --dont-wait --dont-start-it --one-time-build --enable-plugin-otel --disable-plugin-go --disable-plugin-scripts
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply .github/workflows/build.yml:1332]: The three macOS jobs did not provide effective Go/scripts coverage before this change: the previous run disabled both because the required Go toolchain was unavailable. Explicitly disabling them on all three runners makes that existing CI partition deterministic and prevents a future runner-image update from silently adding three full Go builds. The OS_MACOS ndsudo target is still compiled and installed by all three jobs, while its external smartctl integration is intentionally exercised on one representative shard to avoid multiplying Homebrew/test cost.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -1316,6 +1316,10 @@ jobs:
         if: needs.file-check.outputs.run == 'true'
         run: |
           brew install bash
+      - name: Install ndsudo test dependencies
+        if: needs.file-check.outputs.run == 'true' && matrix.name == 'macos-15-M1'
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply .github/workflows/build.yml:1320]: The one-shard runtime scope is intentional. All three macOS versions compile and install the Apple-specific helper, while the exercised policy is bounded string validation plus fixed argv execution and does not call a version-specific macOS API. macos-15-M1 covers the setuid transition and real smartctl argv contract; installing jq/smartmontools and repeating the same external integration on every shard would add the CI multiplication this change is designed to avoid. The remaining cross-version risk is accepted as compile coverage on 14/26 plus runtime coverage on 15.

**Review by stelfrag — APPROVED** [review]
